File: Source/App/Client/src/types/watchRound.ts
export interface WatchRoundQueueItem {
  id: string;
  order: number;
  isRecommendation: boolean;
  mediaId: string | null;
  recommendationId: string | null;
  title: string;
  mediaType: string;
  year: number | null;
  posterUrl: string | null;
  imdbRating: number | null;
  description: string | null;
  genres: string[];
}

export interface ParticipantDecision {
  queueItemId: string;
  approved: boolean;
}

export interface WatchRoundParticipant {
  id: string;
  name: string;
  joinedAt: string;
  isFinished: boolean;
  finishedAt: string | null;
  decisions: ParticipantDecision[];
  approvedItems: WatchRoundQueueItem[];
}

export interface WatchRoundSummary {
  id: string;
  createdAt: string;
  status: 'Active' | 'Finished';
  participantCount: number;
  participantNames: string[];
  queueLength: number;
  mutuallyApprovedCount: number;
  finishedAt: string | null;
}

export interface WatchRoundDetail {
  id: string;
  createdAt: string;
  status: 'Active' | 'Finished';
  includeRecommendations: boolean;
  includeTvShows: boolean;
  minImdbRating: number | null;
  allowedGenres: string[];
  queue: WatchRoundQueueItem[];
  participants: WatchRoundParticipant[];
  mutuallyApprovedItems: WatchRoundQueueItem[];
  mutuallyApprovedCount: number;
  finishedAt: string | null;
}

export interface CreateWatchRoundRequest {
  includeRecommendations: boolean;
  includeTvShows: boolean;
  minImdbRating?: number;
  allowedGenres: string[];
}

export interface CreateWatchRoundResult {
  roundId: string;
  sharePath: string;
  round: WatchRoundDetail;
}

export interface JoinWatchRoundResult {
  participantId: string;
  round: WatchRoundDetail;
}

export function getStoredParticipantId(roundId: string): string | null {
  return localStorage.getItem(`watch-round-participant:${roundId}`);
}

export function storeParticipantId(roundId: string, participantId: string): void {
  localStorage.setItem(`watch-round-participant:${roundId}`, participantId);
}

export function getNextQueueItem(
  round: WatchRoundDetail,
  participant: WatchRoundParticipant,
): WatchRoundQueueItem | null {
  const decided = new Set(participant.decisions.map((d) => d.queueItemId));
  return round.queue.find((item) => !decided.has(item.id)) ?? null;
}

File: Source/App/Api/Models/WatchRound.cs
namespace MoviesAndTVShowsToDo.Api.Models;

public enum WatchRoundStatus
{
    Active,
    Finished
}

public class WatchRoundSettings
{
    public bool IncludeRecommendations { get; set; }
    public bool IncludeTvShows { get; set; } = true;
    public double? MinImdbRating { get; set; }
    public List<string> AllowedGenres { get; set; } = [];
}

public class WatchRoundQueueItem
{
    public Guid Id { get; set; }
    public int Order { get; set; }
    public bool IsRecommendation { get; set; }
    public Guid? MediaId { get; set; }
    public Guid? RecommendationId { get; set; }
    public string Title { get; set; } = string.Empty;
    public MediaType Type { get; set; }
    public int? Year { get; set; }
    public string? PosterUrl { get; set; }
    public double? ImdbRating { get; set; }
    public string? Description { get; set; }
    public List<string> Genres { get; set; } = [];
}

public class ParticipantDecision
{
    public Guid QueueItemId { get; set; }
    public bool Approved { get; set; }
}

public class WatchRoundParticipant
{
    public Guid Id { get; set; }
    public string Name { get; set; } = string.Empty;
    public DateTimeOffset JoinedAt { get; set; }
    public bool IsFinished { get; set; }
    public DateTimeOffset? FinishedAt { get; set; }
    public List<ParticipantDecision> Decisions { get; set; } = [];
}

public class WatchRound
{
    public Guid Id { get; set; }
    public DateTimeOffset CreatedAt { get; set; }
    public WatchRoundStatus Status { get; set; }
    public WatchRoundSettings Settings { get; set; } = new();
    public List<WatchRoundQueueItem> Queue { get; set; } = [];
    public List<WatchRoundParticipant> Participants { get; set; } = [];
    public DateTimeOffset? FinishedAt { get; set; }
}
